### PR TITLE
[WX-1675] Record cloud quota delay to GroupMetrics table

### PR DESCRIPTION
**Pull request by salonishah11**
Jira: https://broadworkbench.atlassian.net/browse/WX-1675

### Description

- Records quota exhaustion for a hog group to the new `GroupMetrics` table when a job runs into `AwaitingCloudQuota` state
- This PR adds a new singleton Actor called `GroupMetricsActor` which takes in currently `RecordGroupQuotaExhaustion` message to record the group that ran into quota exhaustion into the new table

Example screenshot:
Screenshot 2024-08-20 at 4 27 49 PM

#### `CHANGELOG.md`
 - [ ] I updated `CHANGELOG.md` in this PR
 - [ ] I assert that this change shouldn't be included in `CHANGELOG.md` because it doesn't impact community users

#### Terra Release Notes
 - [ ] I added a suggested release notes entry in this Jira ticket
 - [x] I assert that this change doesn't need Jira release notes because it doesn't impact Terra users

### CODE REVIEW

File: backend/src/main/scala/cromwell/backend/standard/StandardAsyncExecutionActor.scala [hunk 1]
```diff
@@ -1057,6 +1058,12 @@ trait StandardAsyncExecutionActor
     */
   def retryEvaluateOutputs(exception: Exception): Boolean = false
 
+  /**
+   * Checks if the job has run into any cloud quota exhaustion and records it to GroupMetrics table
+   * @param runStatus The run status
+   */
+  def checkAndRecordQuotaExhaustion(runStatus: StandardAsyncRunState): Unit = ()
```
Comment (salonishah11) [inline backend/src/main/scala/cromwell/backend/standard/StandardAsyncExecutionActor.scala:1065]: Backend agnostic method and each backend can implement how to detect quota exhaustion. Currently only implemented in `PipelinesApiAsyncBackendJobExecutionActor.scala`

File: engine/src/main/scala/cromwell/server/CromwellRootActor.scala [hunk 2]
```diff
@@ -224,6 +225,9 @@ abstract class CromwellRootActor(terminator: CromwellTerminator,
     "JobExecutionTokenDispenser"
   )
 
+  lazy val groupMetricsActor: ActorRef =
+    context.actorOf(GroupMetricsActor.props(EngineServicesStore.engineDatabaseInterface))
+
```
Comment (salonishah11) [inline engine/src/main/scala/cromwell/server/CromwellRootActor.scala:230]: This is where the singleton GroupMetricsActor is created and wired through other actors to make its way through to backend actor.

File: database/sql/src/main/scala/cromwell/database/slick/GroupMetricsSlickDatabase.scala [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  override def recordGroupMetricsEntry(
+    groupMetricsEntry: GroupMetricsEntry
+  )(implicit ec: ExecutionContext): Future[Unit] = {
+    val action = for {
+      updateCount <- dataAccess
+        .quotaExhaustionForGroupId(groupMetricsEntry.groupId)
+        .update(groupMetricsEntry.quotaExhaustionDetected)
+      _ <- updateCount match {
+        case 0 => dataAccess.groupMetricsEntryIdsAutoInc += groupMetricsEntry
+        case _ => assertUpdateCount("recordGroupMetricsEntry", updateCount, 1)
```
Comment (aednichols) [inline database/sql/src/main/scala/cromwell/database/slick/GroupMetricsSlickDatabase.scala:22]: I hadn't seen this pattern before but I see that it's common in our DB. TIL.

**Review by aednichols — COMMENTED** [review]
I think this branch would be a good mob-time demo. I'm curious to see it in action, but I also don't want to deal with migrating/resetting my local Cromwell DB.

---

Have you thought about where you'd like to place the conversion from "quota timestamp is recent" to `Boolean` "quota is exhausted"?

I was thinking of `quotaExhaustionForGroupId` in the component, but maybe that's too low level and it should wait for the next PR.

**Comment by salonishah11** [comment]
> Have you thought about where you'd like to place the conversion from "quota timestamp is recent" to Boolean "quota is exhausted"?

@aednichols I was thinking probably in the `/database/slick/GroupMetricsSlickDatabase.scala` file or maybe 1 level higher. But yeah that will be included in the next PR which will actually use the values from the table to decide where to allocate new tokens to this group or not.

File: backend/src/test/scala/cromwell/backend/standard/GroupMetricsActorSpec.scala
Comment (salonishah11) [inline backend/src/test/scala/cromwell/backend/standard/GroupMetricsActorSpec.scala:1]: Currently this is only a simple test file that checks the actor receives messages when sent and calls appropriate database method. But in the follow up PR where `GroupMetricsActor` will receive another message related to figuring out whether group is quota exhausted or not, I am expecting that this test file will get more useful tests then.

File: database/sql/src/main/scala/cromwell/database/slick/GroupMetricsSlickDatabase.scala (hunk 3 above)
Comment (salonishah11) [reply database/sql/src/main/scala/cromwell/database/slick/GroupMetricsSlickDatabase.scala:22]: If you had a different pattern in mind to achieve this upsert let me know and I am happy to look into it!

Comment (aednichols) [reply database/sql/src/main/scala/cromwell/database/slick/GroupMetricsSlickDatabase.scala:22]: That was not a leading question, it seems reasonable and I don't know of anything better.

File: backend/src/main/scala/cromwell/backend/standard/GroupMetricsActor.scala [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+import java.time.OffsetDateTime
+
+class GroupMetricsActor(engineDbInterface: EngineSqlDatabase) extends Actor with ActorLogging {
+
+  implicit val ec: MessageDispatcher = context.system.dispatchers.lookup(Dispatcher.EngineDispatcher)
+
+  override def receive: Receive = {
+    case RecordGroupQuotaExhaustion(group) =>
+      log.info(s"Recording quota exhaustion for group '$group'.")
```
Comment (salonishah11) [inline backend/src/main/scala/cromwell/backend/standard/GroupMetricsActor.scala:21]: From discussion - remove this log line

**Review by aednichols — APPROVED** [review]
Nice!

**Review by jgainerdewar — APPROVED** [review]
LGTM once that log line is removed. Thanks for the walkthrough!

**Pull request merged**